cmd/ui.go

```diff
@@ -0,0 +1,154 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2019 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/core/local"
+	"github.com/loadimpact/k6/ui/pb"
+)
+
+// A writer that syncs writes with a mutex and, if the output is a TTY, clears before newlines.
+type consoleWriter struct {
+	Writer io.Writer
+	IsTTY  bool
+	Mutex  *sync.Mutex
+
+	// Used for flicker-free persistent objects like the progressbars
+	PersistentText func()
+}
+
+func (w *consoleWriter) Write(p []byte) (n int, err error) {
+	origLen := len(p)
+	if w.IsTTY {
+		//TODO: check how cross-platform this is...
+		p = bytes.Replace(p, []byte{'\n'}, []byte{'\x1b', '[', '0', 'K', '\n'}, -1)
```

Comment:
Resolved by https://github.com/loadimpact/k6/pull/1007/commits/1a487dac2f03d72c2169dede7dcf8be3f399b8cf? Didn't seem woth the effort to make different methods or functions, because all code usages are slightly different and there wouldn't be a substantial improvement. As I mentioned, the current code is pretty testable.